tests/core/io/test_tcp_server.h

```diff
@@ -0,0 +1,205 @@
+
+	return client_from_server;
+}
+
+TEST_CASE("[TCPServer] Instantiation") {
+	Ref<TCPServer> server;
+	server.instantiate();
+
+	REQUIRE(server.is_valid());
+	CHECK_EQ(false, server->is_listening());
+}
+
+TEST_CASE("[TCPServer] Accept a connection and receive/send data") {
+	Ref<TCPServer> server = create_server(LOCALHOST, PORT);
+	Ref<StreamPeerTCP> client = create_client(LOCALHOST, PORT);
+	Ref<StreamPeerTCP> client_from_server = accept_connection(server);
+
+	REQUIRE_EQ(client->poll(), Error::OK);
+	CHECK_EQ(client->get_status(), StreamPeerTCP::STATUS_CONNECTED);
+
+	// Sending data from client to server.
+	const String hello_world = "Hello World!";
+	client->put_string(hello_world);
+	CHECK_EQ(client_from_server->get_string(), hello_world);
+
+	// Sending data from server to client.
+	const float pi = 3.1415;
+	client_from_server->put_float(pi);
+	CHECK_EQ(client->get_float(), pi);
+
+	client->disconnect_from_host();
+	server->stop();
+	CHECK_FALSE(server->is_listening());
+}
+
+TEST_CASE("[TCPServer] Handle multiple clients at the same time") {
+	Ref<TCPServer> server = create_server(LOCALHOST, PORT);
+
+	Vector<Ref<StreamPeerTCP>> clients;
+	for (int i = 0; i < 5; i++) {
+		clients.push_back(create_client(LOCALHOST, PORT));
+	}
+
+	Vector<Ref<StreamPeerTCP>> clients_from_server;
+	for (int i = 0; i < clients.size(); i++) {
+		clients_from_server.push_back(accept_connection(server));
+	}
+
+	// Calling poll() to update client status.
+	for (Ref<StreamPeerTCP> &c : clients) {
+		REQUIRE_EQ(c->poll(), Error::OK);
+	}
+
+	// Sending data from each client to server.
+	for (int i = 0; i < clients.size(); i++) {
+		String hello_client = "Hello " + itos(i);
+		clients[i]->put_string(hello_client);
+		CHECK_EQ(clients_from_server[i]->get_string(), hello_client);
+	}
+
+	for (Ref<StreamPeerTCP> &c : clients) {
+		c->disconnect_from_host();
+	}
+	server->stop();
+}
+
+TEST_CASE("[TCPServer] When stopped shouldn't accept new connections") {
+	Ref<TCPServer> server = create_server(LOCALHOST, PORT);
+	Ref<StreamPeerTCP> client = create_client(LOCALHOST, PORT);
+	Ref<StreamPeerTCP> client_from_server = accept_connection(server);
+
+	REQUIRE_EQ(client->poll(), Error::OK);
+	CHECK_EQ(client->get_status(), StreamPeerTCP::STATUS_CONNECTED);
+
+	// Sending data from client to server.
+	const String hello_world = "Hello World!";
+	client->put_string(hello_world);
+	CHECK_EQ(client_from_server->get_string(), hello_world);
+
+	client->disconnect_from_host();
+	server->stop();
+	CHECK_FALSE(server->is_listening());
+
+	Ref<StreamPeerTCP> new_client = create_client(LOCALHOST, PORT);
+
+	// Required to get the connection properly established.
+	const uint64_t time = OS::get_singleton()->get_ticks_usec();
+	while (!server->is_connection_available() && (OS::get_singleton()->get_ticks_usec() - time) < MAX_WAIT_USEC) {
+		OS::get_singleton()->delay_usec(SLEEP_DURATION);
+	}
+
+	CHECK_FALSE(server->is_connection_available());
+
+	Error err = new_client->poll();
```

Comment:
As explained in https://github.com/godotengine/godot/pull/97262#discussion_r1780052490 , https://github.com/godotengine/godot/pull/97262#discussion_r1780967061 , and https://github.com/godotengine/godot/pull/97262#discussion_r1805889374 , a single `poll` without any wait is not the correct thing to do.

I don't know how to explain this further :( .

Reply:
> As explained in https://github.com/godotengine/godot/pull/97262#discussion_r1780052490 , https://github.com/godotengine/godot/pull/97262#discussion_r1780967061 , and https://github.com/godotengine/godot/pull/97262#discussion_r1805889374 , a single poll without any wait is not the correct thing to do.
I don't know how to explain this further :( .

Sorry for the inconveniences. I fixed only the server side on my previous attempts, but I neglected the clients. Hope this time I fixed all of them, in this and the UDP PR.